python/test/unit/fem/test_interpolation.py

```diff
@@ -1025,3 +1026,111 @@ def f_test2(x):
+
+    u_sub_exact = Function(V_sub)
+    u_sub_exact.interpolate(ref_func)
+    atol = 5 * np.finfo(default_scalar_type).resolution
+    np.testing.assert_allclose(u_sub_exact.x.array, u_sub.x.array, atol=atol)
+
+    # Map from sub to parent
+    W = functionspace(mesh, ("DG", 4))
+    w = Function(W)
+
+    cell_imap = mesh.topology.index_map(tdim)
+    num_cells = cell_imap.size_local + cell_imap.num_ghosts
+    parent_to_sub = np.full(num_cells, -1, dtype=np.int32)
+    parent_to_sub[sub_to_parent] = np.arange(len(sub_to_parent))
+
+    # Mapping back needs to be restricted to the subset of cells in the submesh
+    w.interpolate(u_sub_exact, cells=sub_to_parent, cell_map=parent_to_sub)
+
+    w_exact = Function(W)
+    w_exact.interpolate(ref_func, cells=cells)
+
+    np.testing.assert_allclose(w.x.array, w_exact.x.array, atol=atol)
+
+
+def test_submesh_expression_interpolation():
+    """Test interpolation of an expression between a submesh and its parent"""
+    mesh = create_unit_square(MPI.COMM_WORLD, 10, 8, cell_type=CellType.quadrilateral)
+
+    def left_locator(x):
+        return x[0] <= 0.5 + 1e-14
+
+    def ref_func(x):
+        return -3 * x[0] ** 2 + x[1] ** 2
+
+    def grad_ref_func(x):
+        values = np.zeros((2, x.shape[1]), dtype=default_scalar_type)
+        values[0] = -6 * x[0]
+        values[1] = 2 * x[1]
+        return values
+
+    def grad_squared(x):
+        grad = grad_ref_func(x)
+        return grad[0] ** 2 + grad[1] ** 2
+
+    tdim = mesh.topology.dim
+    cells = locate_entities(mesh, tdim, left_locator)
+    submesh, sub_to_parent, _, _ = create_submesh(mesh, tdim, cells)
+
+    V = functionspace(mesh, ("Lagrange", 2))
+    u = Function(V)
+    u.interpolate(ref_func)
+
+    V_sub = functionspace(submesh, ("N2curl", 1))
+    u_sub = Function(V_sub)
+
+    parent_expr = Expression(ufl.grad(u), V_sub.element.interpolation_points())
+
+    # Map from parent to sub mesh
+
+    u_sub.interpolate(parent_expr, expr_mesh=mesh, cell_map=sub_to_parent)
+
+    u_sub_exact = Function(V_sub)
+    u_sub_exact.interpolate(grad_ref_func)
+    atol = 5 * np.finfo(default_scalar_type).resolution
+    np.testing.assert_allclose(u_sub_exact.x.array, u_sub.x.array, atol=atol)
+
+    # Map from sub to parent
+    W = functionspace(mesh, ("DQ", 2))
+    w = Function(W)
+
+    cell_imap = mesh.topology.index_map(tdim)
+    num_cells = cell_imap.size_local + cell_imap.num_ghosts
+    parent_to_sub = np.full(num_cells, -1, dtype=np.int32)
+    parent_to_sub[sub_to_parent] = np.arange(len(sub_to_parent))
+
+    sub_expr = Expression(ufl.dot(u_sub_exact, u_sub_exact), W.element.interpolation_points())
+
+    # Mapping back needs to be restricted to the subset of cells in the submesh
+    w.interpolate(sub_expr, cells=sub_to_parent, expr_mesh=submesh, cell_map=parent_to_sub)
+
+    w_exact = Function(W)
+    w_exact.interpolate(grad_squared, cells=cells)
+    np.testing.assert_allclose(w.x.array, w_exact.x.array, atol=5e-4)
```

Comment:
Tolerance seems quite high. Why?

Reply:
Honestly, I do not know.

Reply:
Stricter tolerances fail single precision checks: 
https://github.com/FEniCS/dolfinx/actions/runs/8787873760/job/24114119650?pr=3114
Note that the difference is very small:
```bash
E           Mismatched elements: 79 / 720 (11%)
E           Max absolute difference: 1.8119812e-05
E           Max relative difference: 0.52334005
E            x: array([4.024558e-16, 6.250000e-02, 3.600000e-01, 4.224998e-01,
E                  1.562500e-02, 8.999999e-02, 1.524999e-01, 3.756249e-01,
E                  1.056250e-01, 6.249999e-02, 2.500000e-01, 4.224999e-01,...
E            y: array([8.443247e-16, 6.250000e-02, 3.600000e-01, 4.225002e-01,
E                  1.562500e-02, 9.000000e-02, 1.525001e-01, 3.756251e-01,
E                  1.056250e-01, 6.249999e-02, 2.500000e-01, 4.225000e-01,...

```
Seems like there is some floating point issue in the evaluation?

Reply:
There seems to be some floating error accumulation over multiple steps here. I have revised the test to avoid floating point error accummulation